lmcache/experimental/lookup_server/redis_server.py

```diff
@@ -28,31 +47,128 @@ def __init__(self, config: LMCacheEngineConfig):
+            except asyncio.TimeoutError:
+                break
+
+    def _add_to_pipeline(self, pipe: redis.client.Pipeline, item: _WorkItem):
+        """
+        Add a request to a redis pipeline (batch of requests).
+
+        :param redis.client.Pipeline pipe: The redis pipeline to add to.
+
+        :param _WorkItem item: The item defining the request we want to add.
+        """
+        match item.op:
+            case _Op.INSERT:
+                pipe.hset(self._get_indexing_key(item.key),
+                          self.distributed_url,
+                          self._get_indexing_metadata(item.key))
+            case _Op.DELETE:
+                pipe.hdel(self._get_indexing_key(item.key),
+                          self.distributed_url)
+            case _Op.LOOKUP:
+                pipe.hgetall(self._get_indexing_key(item.key))
+
+    async def _requester(self):
+        """
+        _requester batches requests to be sent to redis
+        """
+        while True:
+            items = []
+            with self.connection.pipeline() as pipe:
+                async for item in self._get_batch():
+                    self._add_to_pipeline(pipe, item)
+                    items.append(item)
+                logger.debug(f"Sending a batch of {len(items)} requests")
+                results = pipe.execute()
+                logger.debug("Batch results are ready")
+
+            for item, result in zip(items, results):
+                item.fut.set_result(result)
+
+    def _run_loop(self):
+        asyncio.set_event_loop(self.loop)
+        self.loop.run_forever()
+
+    def close(self):
+        if self.loop.is_running():
+            self.loop.call_soon_threadsafe(self.loop.stop)
+        if self.thread.is_alive():
+            self.thread.join()
+        self.connection.close()
+
+    def _get_indexing_key(self, key: CacheEngineKey) -> str:
         return f"{key.model_name}@{key.chunk_hash}"
 
-    def _get_indexing_metadata(self, key: CacheEngineKey):
+    def _get_indexing_metadata(self, key: CacheEngineKey) -> str:
         return f"{key.fmt}@{key.world_size}@{key.worker_id}@{time.time()}"
 
-    def _extract_cache_keys_componenets(self, indexing_metadata: str):
-        return indexing_metadata.split("@")[:-1]
+    def _extract_cache_keys_componenets(self, metadata: str) -> List[str]:
+        return metadata.split("@")[:-1]
 
     def lookup(self, key: CacheEngineKey) -> Optional[Tuple[str, int]]:
         """
         Perform lookup in the lookup server.
         """
         logger.debug("Call to lookup in lookup server")
-        result = self.connection.hgetall(self._get_indexing_key(key))
-        logger.debug(f"KV cache lives on {result}")
+        fut = self.loop.create_future()
+        item = _WorkItem(key, _Op.LOOKUP, fut)
+        asyncio.run_coroutine_threadsafe(
+            self.queue.put((_WorkPriority.LOOKUP, item)), self.loop)
+
+        cfut = asyncio.run_coroutine_threadsafe(_wait_for_future(fut),
```

Comment:
Can we configure the timeout here? If an answer does not come quickly enough - better not block the caller.